### PR TITLE
Provide AT Parser thread safety

### PR DESCRIPTION
**Pull request by VeijoPesonen**
Makes access to AT Parser mutually exclusive

### CODE REVIEW

File: ESP8266/ESP8266.cpp [hunk 1]
```diff
@@ -265,6 +321,10 @@ void ESP8266::_packet_handler()
 int32_t ESP8266::recv(int id, void *data, uint32_t amount)
 {
     while (true) {
+        _smutex.lock();
+        // Poll for inbound packets
+        _parser.process_oob();
```
Comment (kjbracey-arm) [inline ESP8266/ESP8266.cpp]: Why did this extra process call need to be added? Not saying it's wrong, just that it's changing the code flow a bit, and it makes me a little nervous.

Comment (SeppoTakalo) [reply ESP8266/ESP8266.cpp]: You are also ignoring errors from this `process_oob()` call.
Then, end of the loop, the same function is called but errors are checked.

You need to decide whether it needs to be called beginning of the loop, or at the end of the loop. And move it there (with error checks).

File: ESP8266/ESP8266.h [hunk 2]
```diff
@@ -218,6 +219,7 @@ class ESP8266
 private:
     UARTSerial _serial;
     ATCmdParser _parser;
+    Mutex _smutex; // One at a time must access the serial line
```
Comment (SeppoTakalo) [inline ESP8266/ESP8266.h]: "Must" is a bit wrong, its more like "only one should access the ..."
Change it to something like "// Protect serial port access"

File: ESP8266/ESP8266.cpp (hunk 1 above)
Comment (VeijoPesonen) [reply ESP8266/ESP8266.cpp]: Beginning it is and while-loop will be removed.

**Comment by VeijoPesonen** [comment]
All the findings addressed. Lets see is there are still some new ones left.

**Pull request merged**